pep-9999.rst

```diff
@@ -0,0 +1,439 @@
+want control over how the output TOML ends up being formatted (how to format
+strings, when to inline arrays or tables, how much to indent, whether to reorder
+contents, etc), they will not be served well by the proposed API.
+
+The standard library does not need to do everything and if we feel that most
+users are better served by more powerful third party write APIs, exclusion is
+acceptable (and could be revisited later).
+
+However, users will likely expect a write API to be available for consistency.
+Empirically, writing TOML seems useful, e.g. ``toml.dump`` is used about 30% as
+often as ``toml.load`` based on https://grep.app
+
+Even a simple API is capable of serving common use cases, such as testing code
+that loads TOML or writing simple or boilerplate TOML.
+TODO: about 1/5 uses of ``toml.dump[s]`` are in tests, estimate other simple use cases
+
+If we keep feature set narrow, a write API shouldn't be too much additional
+burden. The proposed implementation is about 200 lines of code.
+
+Finally, an open issue is whether we're able to re-use the ``toml`` package name.
+If so, having a basic write API could reduce disruption for affected users.
+
+
+Assorted API details
+--------------------
+
+Controlling the type of mappings returned by ``tomllib.load[s]``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This would work similarly to common uses for the ``object_hook`` argument in
+``json.load[s]``.
+
+Such an argument is not necessary for the core use cases outlined in the
+motivation section. The absence of this can be pretty easily worked around using
+a wrapper class or transformer function. Finally, support could be added later
+in a backward compatible way.
+
+The ``toml`` library on PyPI supports this feature using the ``_dict`` argument.
+There are several uses of this on https://grep.app, however, almost all of them
+were passing ``_dict=OrderedDict``, which should no longer be necessary post
+Python 3.7. There were two instances of legitimate use: in one case, a custom
+class was passed for friendlier KeyErrors, in another case, the custom class had
+several additional lookup and mutation methods (e.g. to help resolve dotted
+keys).
+
+Types accepted by the first argument of ``tomllib.load``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``toml`` library on PyPI allows passing path-like objects (and lists of
+path-like objects, reading the first path that exists). Doing this would be
+inconsistent with ``json.load``, ``pickle.load``, etc. If we agree consistency
+with other stdlib modules is desirable, this is somewhat out of scope for this
+PEP. This can easily and perhaps more explicitly be worked around in user code.
+
+The proposed API takes a ``SupportsRead[bytes]``, while ``toml.load`` takes a
+``SupportsRead[str]`` and ``json.load`` takes ``SupportsRead[str | bytes]``.
+While slightly opinionated, this was changed in ``tomli`` v1.2 to a)
```

Comment:
Deprecation of `SupportsRead[str]` happened in v1.2.0 but support was only removed in v2.0.0.

Regarding the use of the word "opinionated": I'm obviously biased here :smile: but I think the change was "correct" not "opinionated". JSON is compatible with text file objects because it allows different file encodings and newline sequences, TOML does not.

Reply:
Yeah, you're right. This was copy-pasted from my more colourfully worded gist. The history here isn't really relevant, I'll just specify the reasons for the current behaviour.

Reply:
Fixed in https://github.com/hauntsaninja/peps/pull/1/commits/956d3bb43f0b8822ad64ad02f46f3b8b51a2f0a0